### ISSUE TITLE
Problems with tacacs+ on Cisco

### ISSUE THREAD

**Issue by cla777** [networktocode/ntc-ansible#123]
I have an Ansible Playbook that does a ntc_save_config on two test switches and using ntc_get_facts saves the date-stamped config files locally on my Ansible server.  The playbook works with the the following AAA config on the switches (local auth):
aaa new-model
aaa authentication login default local
aaa authentication login admin group tacacs+ enable
aaa authorization exec default local none
!
line vty 0 4
 privilege level 15
 logging synchronous
 length 0
login authentication default
 transport input telnet ssh

But I need tacacs+, which the Playbook fails with the following AAA config (tacacs+ auth):
aaa new-model
aaa authentication login default local
aaa authentication login admin group tacacs+ enable
aaa authorization exec default local none
!
line vty 0 4
 privilege level 15
 logging synchronous
 length 0
login authentication admin
 transport input telnet ssh

Fails with "netmiko.ssh_exception.NetMikoAuthenticationException: Authentication failure: unable to connect cisco_ios <ip addr>"

To summarize: ntc-ansible works with a switch configured for local auth, but fails with a switch configured for tacacs+.   I need it to work with tacacs+.  thanks

**Comment by jedelman8** [comment]
I thought I commented on this, but maybe not.  @ktbyers, I know you've seen this before, any recommendations on this?

**Comment by ktbyers** [comment]
Yes, @cla777 and I have been talking in a separate thread on this issue in Netmiko. Let me review that.

**Comment by ktbyers** [comment]
@cla777 Please post comments here so we have a record of them:

You mentioned the following:

```
A bit of clarification:  looks like ntc-ansible has a problem with TACACS+, as the ntc_get_facts module also fails.

Here is a partial of the error I receive when running the playbook with the -vvv option:

netmiko.ssh_exception.NetMikoAuthenticationException:Authentication failure: unable to connect cisco_ios <ip addr> authentication failed.
```

I think we need to get it out of Ansible and do a simple Netmiko authentication test using Python.

```python
#!/usr/bin/env python
from netmiko import ConnectHandler
from getpass import getpass

ip_addr = raw_input("Enter IP Address: ")

device = {
    'device_type': 'cisco_ios',
    'ip': ip_addr,
    'username': 'pyclass',
    'password': getpass(),
    'port': 22,
}

net_connect = ConnectHandler(**device)
output = net_connect.send_command_expect("show version")

print
print '#' * 50
print output
print '#' * 50
print
```

What happens when you execute this?

**Comment by bordeltabernacle** [comment]
I ran into a similar issue with netmiko recently trying to connect to devices with tacacs authentication.
I had to add in an expect like part to my script to enter the password to move into enable mode.

This is the workaround I added into my script:

```python
  # check if in enable mode
  check = net_connect.check_enable_mode()
  print('==> Enable mode? ', check)
  # if not, enter enable mode
  if not check:
    enter_enable = net_connect.send_command('enable')
    print(enter_enable)
    enter_password = net_connect.send_command(PASSWORD)
    print(enter_password)
```

Not sure how relevant this is? :smiley:

**Comment by jedelman8** [comment]
@ktbyers - did you add fixes for this in netmiko?

**Comment by ktbyers** [comment]
@jedelman8  I think the Netmiko TACACS+ is possible fixed, but I don't have a good way to test it. I am definitely interested in hearing if people are still having issues with it.

Need to make sure they are using Netmiko >= 1.0.0 however.

**Comment by itdependsnetworks** [comment]
Seems to be fixed already, going to close.
